source/logic.rst

```diff
@@ -0,0 +1,1223 @@
+
+    import data.real.basic
+
+    -- BEGIN
+    example (a b : ℝ) : abs a = abs (a - b + b) :=
+    by  { congr, ring }
+    -- END
+
+Here the ``congr`` tactic peels off the ``abs`` on each side,
+leaving us to prove ``a = a - b + b``.
+
+Finally, the ``convert`` tactic is used to apply a theorem
+to a goal when the conclusion of the theorem doesn't quite match.
+For example, suppose we want to prove ``a < a * a`` from ``1 < a``.
+A theorem in the library, ``mul_lt_mul_right``,
+will let us prove ``1 * a < a * a``.
+One possibility is to work backwards and rewrite the goal
+so that it has that form.
+Instead, the ``convert`` tactic lets us apply the theorem
+as it is,
+and leaves us with the task of proving the equations that
+are needed to make the goal match.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    -- BEGIN
+    example {a : ℝ} (h : 1 < a) : a < a * a :=
+    begin
+      convert (mul_lt_mul_right _).2 h,
+      { rw [one_mul] },
+      exact lt_trans zero_lt_one h
+    end
+    -- END
+
+This example illustrates another useful trick: when we apply an
+expression with an underscore
+and Lean can't fill it in for us automatically,
+it simply leaves it for us as another goal.
+
+The following shows that any constant sequence :math:`a, a, a, \ldots`
+converges.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    def converges_to (s : ℕ → ℝ) (a : ℝ) :=
+    ∀ ε, 0 < ε → ∃ N, ∀ n, N ≤ n → abs (s n - a) < ε
+
+    variable (a : ℝ)
+
+    -- BEGIN
+    theorem converges_to_const : converges_to (λ x : ℕ, a) a :=
+    begin
+      intros ε εpos,
+      use 0,
+      intros n nge, dsimp,
+      rw [sub_self, abs_zero],
+      apply εpos
+    end
+    -- END
+
+For a more interesting theorem, let's show that if ``s``
+converges to ``a`` and ``t`` converges to ``b``, then
+``λ n, s n + t n`` converges to ``a + b``.
+It is helpful to have a clear pen-and-paper
+proof in mind before you start writing a formal one.
+Given ``ε`` greater than ``0``,
+the idea is to use the hypotheses to obtain an ``Ns``
+such that beyond that point, ``s`` is within ``ε / 2``
+of ``a``,
+and an ``Nt`` such that beyond that point, ``t`` is within
+``ε / 2`` of ``b``.
+Then, whenever ``n`` is greater than or equal to the
+maximum of ``Ns`` and ``Nt``,
+the sequence ``λ n, s n + t n`` should be within ``ε``
+of ``a + b``.
+The following example begins to implement this strategy.
+See if you can finish it off.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    def converges_to (s : ℕ → ℝ) (a : ℝ) :=
+    ∀ ε, 0 < ε → ∃ N, ∀ n, N ≤ n → abs (s n - a) < ε
+
+    -- BEGIN
+    variables {s t : ℕ → ℝ} {a b : ℝ}
+
+    theorem converges_to_add
+      (cs : converges_to s a) (ct : converges_to t b):
+    converges_to (λ n, s n + t n) (a + b) :=
+    begin
+      intros ε εpos, dsimp,
+      have ε2pos : 0 < ε / 2,
+      { linarith },
+      cases cs (ε / 2) ε2pos with Ns hs,
+      cases ct (ε / 2) ε2pos with Nt ht,
+      use max Ns Nt,
```

Comment:
This exercise needs lemmas about `max`.

Reply:
They did some things with `max` in the previous chapter, and at this stage I am hoping that they are starting to get comfortable with tab completion, searching mathlib, etc. But maybe I can add a hint.